tao_compiler/mlir/disc/transforms/fusion_utils_dot_gpu.cc

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022 The BladeDISC Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tensorflow/compiler/mlir/disc/transforms/fusion_utils.h"
+
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/utils/source_emitter.h"
+
+namespace mlir {
+namespace disc_ral {
+
+////////////////////// Dot GPU FusionStrategy Implemenation ////////////
+////////////////////////////////////////////////////////////////////////
+
+bool DotGpuFusionStrategy::isFusible(Operation* op) {
+  // Only scalar const are supported by the fusion codegen engine a.t.m.
+  if (isa<lmhlo::ConstantOp>(op)) {
+    auto constant = cast<lmhlo::ConstantOp>(op);
+    MemRefType type = constant.getOutput().getType().cast<MemRefType>();
+    return (type.getRank() == 0 || constant.getValue().isSplat());
+  }
+
+  return isa<lmhlo::DotGeneralOp>(op) || SourceEmitterCUDA::isSupportedOp(op);
+}
+
+bool DotGpuFusionStrategy::initFusionPattern(ShapeAnalysis& shapeAnalysis,
+                                             FusionPattern& fusion_pattern) {
+  /*
+   * 1. Only one dot is supported per fusion, currently.
+   * 2. Dot is the first op;
+   * 3. Other ops are supported by the CUDA source emitter;
+   * 4. Other ops' effective operands are not the operands of the fusion,
+   *    currently; otherwise this op should be scalar/splat constant.
+   *
+   * The conditions 1 and 4 will be loosed in the future.
+   */
+
+  const auto& op_list = fusion_pattern.getOpList();
+  const auto& operands = fusion_pattern.getOperands();
+  const auto& results = fusion_pattern.getResults();
+
+  SmallVector<Operation*> dot_ops;
+  SmallVector<Operation*> mem_intensive_ops;
+  for (auto op : op_list) {
+    if (isa<lmhlo::DotGeneralOp>(op)) {
+      dot_ops.push_back(op);
+    } else {
+      mem_intensive_ops.push_back(op);
+    }
+  }
+
+  // If it is a single supported elementwise op, it may be fused with dot later.
+  // Do not do multi-op fusion for pure element-wise ops to prevent breaking
+  // possible dot fusion.
+  if (dot_ops.size() == 0 && mem_intensive_ops.size() == 1) {
+    return initFusionPatternBase(shapeAnalysis, fusion_pattern);
+  }
+
+  // Only one dot.
```

Comment:
Just curious, there are two constraints about dot in current strategy, first is only on dot and the second is dot is the first op. The first constraint is checked while the second is ignored. Is dot op can always be the first one in current design?